File: RepoTransBench/source_projects/Python/spotify_pythonflow/tests/test_core_basic.py
import pytest
from pythonflow.core import Graph, Operation

class DummyOp(Operation):
    # minimal operation for testing graph
    def _evaluate(self):
        return 42

def test_graph_enter_exit():
    g = Graph()
    assert getattr(g._globals, 'default_graph', None) is None
    with g:
        assert Graph._globals.default_graph is g
    assert getattr(g._globals, 'default_graph', None) is None

def test_graph_duplicate_enter():
    g = Graph()
    with g:
        with pytest.raises(AssertionError):
            with g:
                pass

def test_graph_normalize_operation_with_instance():
    g = Graph()
    op = DummyOp(name="abc", graph=g)
    g.operations["abc"] = op
    assert g.normalize_operation(op) is op
    # Bad graph
    other = Graph()
    op2 = DummyOp(name="def", graph=other)
    with pytest.raises(RuntimeError):
        g.normalize_operation(op2)

def test_graph_normalize_operation_with_name():
    g = Graph()
    op = DummyOp(name="abc", graph=g)
    g.operations["abc"] = op
    assert g.normalize_operation("abc") is op

def test_graph_normalize_operation_invalid():
    g = Graph()
    with pytest.raises(ValueError):
        g.normalize_operation(123)
    with pytest.raises(KeyError):
        g.normalize_operation("notfound")

def test_graph_normalize_context_and_duplicates():
    g = Graph()
    op = DummyOp(name='x', graph=g)
    g.operations['x'] = op
    # context with string key
    ctx = {'x': 3}
    norm = g.normalize_context(ctx.copy())
    assert op in norm
    # context not a mapping
    with pytest.raises(ValueError):
        g.normalize_context([('x', 3)])
    # context duplicate keys
    ctx_dup = {op: 1, 'x': 2}
    with pytest.raises(ValueError):
        g.normalize_context(ctx_dup.copy())

def test_graph_normalize_context_kwargs():
    g = Graph()
    op = DummyOp(name='x', graph=g)
    g.operations['x'] = op
    norm = g.normalize_context({}, x=99)
    assert op in norm and norm[op] == 99

File: RepoTransBench/target_projects/Python/C++/spotify_pythonflow/tests/test_core_basic.cpp
#include <gtest/gtest.h>
#include "pythonflow/core.h"

class DummyOp : public pythonflow::Operation {
public:
    DummyOp(const std::string &name, pythonflow::Graph* graph) : Operation(name, graph) {}
    int evaluate() override { return 42; }
};

TEST(GraphCoreBasic, EnterExit) {
    pythonflow::Graph g;
    ASSERT_EQ(g.getDefaultGraph(), nullptr);
    {
        pythonflow::Graph::GlobalGraphSetter setter(&g);
        ASSERT_EQ(pythonflow::Graph::getDefaultGraph(), &g);
    }
    ASSERT_EQ(g.getDefaultGraph(), nullptr);
}

TEST(GraphCoreBasic, DuplicateEnter) {
    pythonflow::Graph g;
    pythonflow::Graph::GlobalGraphSetter setter1(&g);
    EXPECT_THROW({
        pythonflow::Graph::GlobalGraphSetter setter2(&g);
    }, std::logic_error /* Replace with real exception type for duplicate enter */);
}

TEST(GraphCoreBasic, NormalizeOperationWithInstance) {
    pythonflow::Graph g;
    auto op = std::make_shared<DummyOp>("abc", &g);
    g.operations["abc"] = op;
    ASSERT_EQ(g.normalizeOperation(op), op);
    pythonflow::Graph other;
    auto op2 = std::make_shared<DummyOp>("def", &other);
    EXPECT_THROW(g.normalizeOperation(op2), std::runtime_error);
}

TEST(GraphCoreBasic, NormalizeOperationWithName) {
    pythonflow::Graph g;
    auto op = std::make_shared<DummyOp>("abc", &g);
    g.operations["abc"] = op;
    ASSERT_EQ(g.normalizeOperation("abc"), op);
}

TEST(GraphCoreBasic, NormalizeOperationInvalid) {
    pythonflow::Graph g;
    EXPECT_THROW(g.normalizeOperation(123), std::invalid_argument);
    EXPECT_THROW(g.normalizeOperation("notfound"), std::out_of_range);
}

TEST(GraphCoreBasic, NormalizeContextAndDuplicates) {
    pythonflow::Graph g;
    auto op = std::make_shared<DummyOp>("x", &g);
    g.operations["x"] = op;
    std::map<std::string, int> ctx { {"x", 3} };
    auto norm = g.normalizeContext(ctx);
    ASSERT_TRUE(norm.count(op));
    // context not a mapping
    EXPECT_THROW(g.normalizeContextVec({{"x", 3}}), std::invalid_argument);
    // duplicate keys
    std::map<pythonflow::OperationPtr, int> ctx_dup;
    ctx_dup[op] = 1;
    // context with op + string key "x"
    EXPECT_THROW(g.normalizeContextMixed(ctx_dup, {{"x", 2}}), std::invalid_argument);
}

TEST(GraphCoreBasic, NormalizeContextKwargs) {
    pythonflow::Graph g;
    auto op = std::make_shared<DummyOp>("x", &g);
    g.operations["x"] = op;
    auto norm = g.normalizeContextWithKwargs({}, {{"x", 99}});
    ASSERT_TRUE(norm.count(op) && norm[op] == 99);
}